client/src/main/scala/org/http4s/client/middleware/DigestClient.scala

```diff
@@ -0,0 +1,180 @@
+          extractChallenge(resp.headers) match {
+            case Some(challenge) =>
+              dispose >> digestAuthHeader(req, challenge) >>= (h => prepareLoop(req.putHeaders(h)))
+            case None =>
+              Resource.make(resp.pure[F])(_ => dispose).pure[F]
+          }
+        case Right((resp, dispose)) => Resource.make(resp.pure[F])(_ => dispose).pure[F]
+        case Left(e) => F.raiseError(e)
+      }
+
+    Client(req => Resource.suspend(prepareLoop(req)))
+  }
+
+  private def extractChallenge(headers: Headers): Option[Challenge] =
+    for {
+      challenge <- headers.get(`WWW-Authenticate`).map(_.value)
+      res <-
+        HttpHeaderParser
+          .WWW_AUTHENTICATE(challenge)
+          .fold(
+            _ => {
+              logger.info(s"Couldn't parse challenge: $challenge")
+              none
+            },
+            v => Some(v.values.head))
+    } yield res
+
+  private def hasDigestAuth(headers: Headers): Boolean =
+    headers
+      .get(Authorization)
+      .exists(_.credentials.authScheme == AuthScheme.Digest)
+
+}
+
+trait NonceCounter[F[_]] {
+
+  /** Returns `cnonce` and next `nc` value for the given nonce */
+  def nc(nonce: String): F[(String, Long)]
+
+  /** Call when a conversation is over, to prevent unbounded growth. */
+  def evict(nonce: String): F[Unit]
+
+  /** Call when a conversation is over, to prevent unbounded growth. */
+  def evictAll: F[Unit]
+}
+
+object NonceCounter {
+  val Max: Long = java.lang.Long.parseLong("FFFFFFFF", 16)
+  def nextCnonce(): String = {
+    val bytes = new Array[Byte](16)
+    Random.nextBytes(bytes)
+    Base64.getEncoder.encodeToString(bytes)
+  }
+  def defaultNonceCounter[F[_]: Concurrent: Timer]: NonceCounter[F] =
+    new SimpleInMemoryStoreNonceCounter[F](1.minute, nextCnonce)
+}
+
+class SimpleInMemoryStoreNonceCounter[F[_]](nonceTtl: FiniteDuration, cnonceF: () => String)(
+    implicit
+    T: Timer[F],
+    C: Concurrent[F])
+    extends NonceCounter[F] {
+  private val cache = TrieMap.empty[String, Ref[F, Entry]]
+  override def nc(nonce: String): F[(String, Long)] =
+    (for {
+      now <- T.clock.realTime(TimeUnit.MILLISECONDS)
+      ref <- cache.getOrElseUpdate(nonce, Ref.unsafe(Entry(now, cnonceF(), 0))).pure[F]
+      res <- ref.updateAndGet { d =>
+        val next = d.copy(usedAt = now, nc = d.nc + 1)
+        if (next.nc > Max) Entry(now, cnonceF(), 1) else next
+      }
+    } yield (res.cnonce, res.nc)) <* C.start(autoCleanIfNotUsed(nonce))
+
+  private def autoCleanIfNotUsed(nonce: String): F[Unit] =
+    for {
+      time <- T.clock.realTime(TimeUnit.MILLISECONDS)
+      _ <- T.sleep(nonceTtl)
+      _ <- cache.get(nonce) match {
+        case Some(ref) => ref.get >>= (r => if (r.usedAt > time) ().pure[F] else evict(nonce))
+        case None => ().pure[F]
+      }
+    } yield ()
+
+  override def evict(nonce: String): F[Unit] =
+    cache.remove(nonce).map(_ => ().pure[F]).getOrElse(().pure[F])
```

Comment:
Instead of `.pure[F]`, we want the cache-modifying operations inside `F.delay`, or else the side effect has already run when we return the `F[Unit]`.  Same with `evictAll`.